src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -697,6 +697,60 @@ private function step_in_body() {
 				$this->state->stack_of_open_elements->pop_until( $tag_name );
 				return true;
 
+			/*
+			 * > A start tag whose tag name is one of: "h1", "h2", "h3", "h4", "h5", "h6"
+			 */
+			case '+H1':
+			case '+H2':
+			case '+H3':
+			case '+H4':
+			case '+H5':
+			case '+H6':
+				if ( $this->state->stack_of_open_elements->has_p_in_button_scope() ) {
+					$this->close_a_p_element();
+				}
+
+				if (
+					in_array(
+						$this->state->stack_of_open_elements->current_node()->node_name,
+						array( 'H1', 'H2', 'H3', 'H4', 'H5', 'H6' ),
+						true
+					)
+				) {
+					// @TODO: Indicate a parse error once it's possible.
+					$this->state->stack_of_open_elements->pop();
+				}
+
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is one of: "h1", "h2", "h3", "h4", "h5", "h6"
+			 */
+			case '-H1':
+			case '-H2':
+			case '-H3':
+			case '-H4':
+			case '-H5':
+			case '-H6':
+				if ( ! $this->state->stack_of_open_elements->has_element_in_scope( '(internal: H1 through H6 - do not use)' ) ) {
```

Comment:
As an alternative to the string literal, we could spell out the whole list

```suggestion
				if (
					! $this->state->stack_of_open_elements->has_element_in_scope( 'H1' ) &&
					! $this->state->stack_of_open_elements->has_element_in_scope( 'H2' ) &&
					/* etc */

				) {
```

Reply:
thank you. as we discussed, I think this is going to be very difficult to do in practice for now, specifically because of the `pop_until()` behavior. using each element individually requires looping in a way that we don't need to do. I'd prefer we leave this a bit awkward internally for now until we have a better solution that doesn't tradeoff efficiency to get there.